src/pd/retry.rs

```diff
@@ -131,99 +109,153 @@ impl fmt::Debug for RetryClient {
     }
 }
 
-fn retry_request<Resp, Func, RespFuture>(
-    client: Arc<RetryClient>,
-    func: Func,
-) -> RetryRequest<impl Future<Output = Result<Resp>>>
-where
-    Resp: Send + 'static,
-    Func: Fn(&Cluster) -> RespFuture + Send + 'static,
-    RespFuture: Future<Output = Result<Resp>> + Send + 'static,
-{
-    let mut req = Request::new(func, client);
-    RetryRequest {
-        reconnect_count: LEADER_CHANGE_RETRY,
-        future: req
-            .reconnect_if_needed()
-            .map_err(|_| internal_err!("failed to reconnect"))
-            .and_then(move |_| req.send_and_receive()),
-    }
+// A node-like thing that can be connected to.
+trait Reconnect {
+    type Cl;
+    fn reconnect(&self, interval: u64) -> Result<()>;
+    fn with_cluster<T, F: Fn(&Self::Cl) -> T>(&self, f: F) -> T;
 }
 
-/// A future which will retry a request up to `reconnect_count` times or until it
-/// succeeds.
-struct RetryRequest<Fut> {
-    reconnect_count: usize,
-    future: Fut,
-}
+impl Reconnect for RetryClient<Cluster> {
+    type Cl = Cluster;
 
-struct Request<Func> {
-    // We keep track of requests sent and after `MAX_REQUEST_COUNT` we reconnect.
-    request_sent: usize,
-
-    client: Arc<RetryClient>,
+    fn reconnect(&self, interval: u64) -> Result<()> {
+        if let Some(cluster) =
+            self.connection
+                .reconnect(&self.cluster.read().unwrap(), interval, self.timeout)?
+        {
+            *self.cluster.write().unwrap() = cluster;
+        }
+        Ok(())
+    }
 
-    // A function which makes an async request.
-    func: Func,
+    fn with_cluster<T, F: Fn(&Cluster) -> T>(&self, f: F) -> T {
+        f(&self.cluster.read().unwrap())
+    }
 }
 
-impl<Resp, Fut> Future for RetryRequest<Fut>
+async fn retry_request<Rc, Resp, Func, RespFuture>(client: Arc<Rc>, func: Func) -> Result<Resp>
 where
+    Rc: Reconnect,
     Resp: Send + 'static,
-    Fut: Future<Output = Result<Resp>> + Send + 'static,
+    Func: Fn(&Rc::Cl) -> RespFuture,
+    RespFuture: Future<Output = Result<Resp>> + Send + 'static,
 {
-    type Output = Result<Resp>;
-
-    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<Resp>> {
-        unsafe {
-            let this = Pin::get_unchecked_mut(self);
-            if this.reconnect_count == 0 {
-                return Poll::Ready(Err(internal_err!("failed to send request")));
+    let mut last_err = Ok(());
+    for _ in 0..LEADER_CHANGE_RETRY {
+        // Reconnect if necessary.
+        let mut reconnect_count = MAX_REQUEST_COUNT;
+        while let Err(e) = client.reconnect(RECONNECT_INTERVAL_SEC) {
```

Comment:
Looks like we always reconnect on the first try. We should reconnect only if there is an error.

Reply:
addressed